calico_node/tests/st/utils/workload.py

```diff
@@ -51,20 +54,72 @@ def __init__(self, host, name, image="busybox", network="bridge", ip=None, label
         """
         self.host = host
         self.name = name
+        self.network = network
 
         lbl_args = ""
         for label in labels:
             lbl_args += " --label %s" % (label)
 
         ip_option = ("--ip %s" % ip) if ip else ""
-        command = "docker run -tid --name %s --net %s %s %s %s" % \
-                  (name, network, lbl_args, ip_option, image)
+        command = "docker run -tid --name %s --net=none %s %s" % \
+                  (name, lbl_args, image)
+        # Change to --net none here, and use Calico CNI instead.
         docker_run_wl = partial(host.execute, command)
         retry_until_success(docker_run_wl)
-        self.ip = host.execute(
-            "docker inspect --format "
-            "'{{.NetworkSettings.Networks.%s.IPAddress}}' %s"
-            % (network, name))
+        #self.ip = host.execute(
+        #    "docker inspect --format "
+        #    "'{{.NetworkSettings.Networks.%s.IPAddress}}' %s"
+        #    % (network, name))
+        self.run_cni("ADD", ip=ip)
+
+    def run_cni(self, add_or_del, ip=None):
+        adding = (add_or_del == "ADD")
+        workload_pid = self.host.execute("docker inspect --format '{{.State.Pid}}' %s" % self.name)
+        container_id = self.host.execute("docker inspect --format '{{.Id}}' %s" % self.name)
+        ip_json = (",\"args\":{\"ip\":\"%s\"}" % ip) if (ip and adding) else ""
+        ip_args = ("CNI_ARGS=IP=%s " % ip) if (ip and adding) else ""
+        etcd_args = "ETCD_ENDPOINTS=http://%s:2379 " % get_ip()
+        if ETCD_SCHEME == "https":
+            etcd_args = ("ETCD_ENDPOINTS=https://%s:2379 " % ETCD_HOSTNAME_SSL +
+                         "ETCD_CA_CERT_FILE=%s " % ETCD_CA +
+                         "ETCD_CERT_FILE=%s " % ETCD_CERT +
+                         "ETCD_KEY_FILE=%s " % ETCD_KEY)
+        command = ("echo '{" +
+                   "\"name\":\"%s\"," % self.network +
+                   "\"type\":\"calico-cni-plugin\"," +
+                   "\"ipam\":{\"type\":\"calico-ipam-plugin\"%s}" % ip_json +
+                   "}' | " +
+                   etcd_args +
+                   "CNI_COMMAND=%s " % add_or_del +
+                   "CNI_CONTAINERID=%s " % container_id +
+                   "CNI_NETNS=/proc/%s/ns/net " % workload_pid +
+                   "CNI_IFNAME=eth0 " +
+                   "CNI_PATH=/code " +
+                   ip_args +
+                   "/code/calico-cni-plugin")
+        output = self.host.execute(command)
+
+        if adding:
+            # The CNI plugin writes its logging to stderr and its JSON output -
+            # including the IP address that we need - to stdout, but
```

Comment:
Is the problem that the tests run outside of the docker-in-docker host container so we can't distinguish the output returned from docker exec?

I think this logic is probably sound enough. Another option might be to query the workloadendpoint created by the CNI plugin and use the address from that, but I suspect that's more work than it's worth.  Up to you.

Reply:
doesn't really help now, but would it be worth adding a logging option to CNI to prefix log lines with something nice an searchable?

Reply:
Yes, that is the problem. Should I add something to the comment about that?

Reply:
@robbrockbank There already is a "Calico CNI" prefix on the lines that the plugin emits itself, and for the non-SSL tests it works fine to use that as the heuristic. But with SSL there are other emitted lines - something about ignoring the certificate subject name - which I think come from some library, and those are not prefixed.

Reply:
ah gotcha.  grrr that's annoying.